builder/builder.go

```diff
@@ -836,6 +839,58 @@ func (b *Builder) ListBundles(listType listType, tree bool) error {
 	return nil
 }
 
+var validEditors = map[string]string{
+	"joe":   "/usr/bin/joe",
+	"emacs": "/usr/bin/emacs",
+	"ctags": "/usr/bin/ctags",
+	"nano":  "/usr/bin/nano",
+	"vim":   "/usr/bin/vim",
+}
+
+//EditBundles prints out a bundle list in either a flat list or tree view
+func (b *Builder) EditBundles(bundles []string, noEdit bool, editor string, add bool) error {
+	if editor != "" {
+		b.BundleEditor = editor
+	}
+	editorCmd, exists := validEditors[b.BundleEditor]
```

Comment:
This is actually exactly the original way I had this implemented -- if the user sets (or passes in) the path to the editor, use that, else try to look up $VISUAL, then $EDITOR, then fall back to nano.

The reason I went away from this is that, while this flexibility is awesome, basic security policy says to explicitly _not_ "trust your users!".  Whatever you configure here is going to just be passed off to an `exec`, effectively. Unlike `git`, which lets you configure whatever editor you want for merge conflicts, `mixer` is run via `sudo`. Yes, _these_ commands require no elevated privileges, but when so many of the others do, users will absolutely run them with `sudo` either out of a) laziness, b) misunderstanding, or c) because the files they're editing are owned by root because of some other operation.

As far as the list not including xemacs or gvim, I just pulled the list from the packages included in `editors` bundle -- if those packages include other variants I left out, I'm happy to add them.

Reply:
Just to be clear (as I hope is evidenced by the fact that I originally implemented it the way you suggested), I also don't love the idea of limiting this to a small list of editors. I am absolutely open to ways to open this up that don't introduce mixer executing random binaries as root. :)

Reply:
I do understand your concern, but I think you are fixing the problem in the wrong place. We have other of those customizations of what program to call (BuildScript, --packager), that I think is unfeasible to whitelist all of them.

Also, we really don't want the user to create those files with root permissions (if he's keeping track of the repository in git for example).

I think the real way to deal with this is to contain where we really need sudo, and possibly call sudo ourselves when we really need (which also might be controversial, but I think is a fight on the right battle).

For this functionality (which I think is very good btw!), I'd say either:
- Just do what you did before or Icarus is suggesting
- Avoid editing for now and keep in some form the "copy this manifest from upstream" functionality
- Figure out a way to safely "drop privileges" by setuid yourself back to the original user when running this command (kind of the reverse of sudo). I don't think Go handles that very well yet, though. (See their issue 1435).

Reply:
> Personally I would be inclined to not have an option to pass in the editor via a flag at all.

I agree with that one, we can defer the flag until a real need surfaces. The EDITOR and [name]_EDITOR are common enough in the command line.

Reply:
The problem is that even if you restrict the user to only running 'vi' you can fire up vi. You then type ":sh" and enter, and you have a shell running as root! At that point you can run anything you like as root.

sudo is the wrong tool for increasing privileged status. We ought to see exactly what wants 'root', and see if we can eliminate the need as a medium term goal.

Reply:
> We have other of those customizations of what program to call (BuildScript, --packager), that I think is unfeasible to whitelist all of them

`BuildScript` is hard-coded (and not read in from the config), and for `--packager`, you really are running `exec` on whatever the user passes in? I didn't review that part of the new BCB, but I would have definitely flagged that if I had.

> Also, we really don't want the user to create those files with root permissions (if he's keeping track of the repository in git for example).

Agreed.

> I think the real way to deal with this is to contain where we really need sudo

:+1: 

> Figure out a way to safely "drop privileges" by setuid yourself back to the original user when running this command (kind of the reverse of sudo). I don't think Go handles that very well yet, though. (See their issue 1435).

I've done this in other languages before. I'll look into if this is feasible.

Reply:
> We ought to see exactly what wants 'root', and see if we can eliminate the need as a medium term goal.

@icarus-sparry  created #158 with some more data on this. Want to tackle that problem? :-)

Reply:
> BuildScript is hard-coded (and not read in from the config),

That was surprising. I guess it was made that way to be configurable at some point. That should be a constant.

> and for --packager, you really are running exec on whatever the user passes in? I didn't review that part of the new BCB, but I would have definitely flagged that if I had.

Yep, to let the user pick some other dnf (either the system or a compiled one for testing). In that case, we could've retrofit it to be a flag that just says "use dnf" and use whatever is in PATH.

I guess the two examples were not that strong but I stand by the other arguments I gave :-)